# TREC 2019 Precision Medicine

This is an example notebook to get the grip of TREC and check everything's running.
Remember you need to place a **config.json** file under the **trec_utils** directory pointing to the elasticsearch server. 

In [1]:
import json
from json2html import *
from IPython.display import HTML
import pandas

import os, sys
nb_dir = os.path.split(os.getcwd())[0]
if nb_dir not in sys.path:
    sys.path.append(nb_dir)
from trec_utils import utils, running, evaluation

In [2]:
config = utils.load_config()
config

{'ELASTIC': 'http://trecimi.ddns.net:9200',
 'ABSTRACTS': '/abstracts',
 'TRIALS': '/trials'}

Here an example from 2017, where we load and list the topics (with already some preprocessing, like splitting the genes and calculating MeSH age groups).

In [3]:
topics_2017 = utils.get_topics('../input_files/2017.topics.xml')
topics_2017.head(10)

topic                         disease                              gene  \
1       1                     Liposarcoma                CDK4 Amplification   
2       2                    Colon cancer         KRAS (G13D), BRAF (V600E)   
3       3                      Meningioma            NF2 (K322), AKT1(E17K)   
4       4                   Breast cancer  FGFR1 Amplification, PTEN (Q171)   
5       5                        Melanoma     BRAF (V600E), CDKN2A Deletion   
6       6                        Melanoma                       NRAS (Q61K)   
7       7                     Lung cancer                      EGFR (L858R)   
8       8                     Lung cancer        EML4-ALK Fusion transcript   
9       9  Gastrointestinal stromal tumor         KIT Exon 9 (A502_Y503dup)   
10     10             Lung adenocarcinoma                       KRAS (G12C)   

                         gene1             gene2 gene3     sex age age_group  
1           CDK4 Amplification                            male  38     adult  
2                  KRAS (G13D)      BRAF (V600E)          male  52      aged  
3                   NF2 (K322)        AKT1(E17K)        female  45      aged  
4          FGFR1 Amplification       PTEN (Q171)        female  67      aged  
5                 BRAF (V600E)   CDKN2A Deletion        female  45      aged  
6                  NRAS (Q61K)                            male  55      aged  
7                 EGFR (L858R)                          female  50      aged  
8   EML4-ALK Fusion transcript                            male  52      aged  
9    KIT Exon 9 (A502_Y503dup)                          female  49      aged  
10                 KRAS (G12C)                          female  61      aged

The aim of the competition is to produce a file with relevant results for all topics (maximum: 1000 per topic). That means, to find the top 1000 most relevant documents (could be less) from PubMed for each topic. This is called a **run in trec_eval format**. Each group con submit up to 5 runs for each subtask (5 for abstracts, 5 for clinical trials). Below is an example we submitted in 2017. **Note than in 2019 there will be 3 new columns after the run name to add treatments**.

In [4]:
example_run = utils.from_trec_run_file('../../archived/2017/submitted/mugpubbase.trec_results')
example_run.sort_values(by=['TOPIC_NO', 'RANK']).head(10)

TOPIC_NO  Q0        ID  RANK      SCORE    RUN_NAME
16249         1  Q0  25028469     1  65.013748  mugpubbase
16250         1  Q0  23852861     2  64.275940  mugpubbase
16251         1  Q0  26643872     3  60.410633  mugpubbase
16252         1  Q0  26336885     4  59.411972  mugpubbase
16253         1  Q0  25121597     5  58.291725  mugpubbase
16254         1  Q0  11505267     6  57.617733  mugpubbase
16255         1  Q0  24065146     7  55.737698  mugpubbase
16256         1  Q0  25059573     8  55.404232  mugpubbase
16257         1  Q0  26642065     9  54.019871  mugpubbase
16258         1  Q0  23569312    10  52.573662  mugpubbase

This means that according to our experiments the most relevant document on pubmed to treat **topic 1** (a 38 male with Liposarcoma and CDK4 amplification) is the article with [PubMed ID 25028469](https://www.ncbi.nlm.nih.gov/pubmed/?term=25028469), and so on.

The file above will be the evaluated against a gold standard (generated with results submitted by participants, so never available for training that year). This file is called a **qrels** file . There are 3 relevance scores: **2=relevant**, **1=partially relevant**, **0=not relevant**.

In [5]:
qrels_2017 = utils.get_qrels('../input_files/2017.abstracts.sorted.qrels')
qrels_2017.head(10)

topic    doc_id  relev
0      1  10755400      2
1      1  10981874      2
2      1  11369052      2
3      1  11505267      2
4      1  12242528      2
5      1  14736600      2
6      1  15154619      2
7      1  15991843      2
8      1  17214366      2
9      1  17895758      2

The ideal situation would be that our submitted file matches exactly the relevance order of the qrels file. That is that all most relevant documents (relev=2) are ranked first, etc. The program [trec_eval](https://github.com/usnistgov/trec_eval) is used by the organization to evaluate the results using a bunch of metrics. This program is written in C and needs to be downloaded and compiled.

Fortunately, there is python package called **pytrec_eval** that provides the same functionality, which I also improved to make it easier to use and focus only on the evaluation measures for the competition.

In [6]:
results, aggregated = evaluation.evaluate(qrels_2017, example_run)
aggregated

{'recall_1000': 0.5224, 'ndcg': 0.4956, 'Rprec': 0.2772, 'P_10': 0.62}

Remember, the final aim is to generate (up to 5) run files like *example_run* using different strategies and submit them to the competition. The usual way is to use the strategies on topics and qrels from 2017 (the only year we have a gold standard available), then re-run them in the topics of the current year and submit them. However, note than in 2019 the competition is slightly different and now [treatments are also requested (optionally)](http://www.trec-cds.org/2019.html).

How to generate the qrels files given our elasticsearch server? Some hints are below and in the abstracts_experiments notebook accompanying this one. The main idea is that each strategy maps to an elasticsearch query and the retrieved results are transformed into a trec_eval file and submitted (also very easily using utilities I wrote). The queries are a collection of strategies from previous years and experimenting.

# FIXME: OLD STUFF FROM 2018

In [7]:
topics_2018=utils.get_topics('../input_files/2018.topics.xml')

## SUBMITTED RUN 1
Baseline query with MeSH age groups and grid-search optimized parameters.
Relevance: 3

Evaluate on 2017 topics and qrels first.

In [8]:
run_params = {
    'run_id':'imi_mug_abs1',
    'query_template':'submitted/imi_mug_abs1.json',
}

run, run_params = running.run(topics_2017, 'ABSTRACTS', run_params)
results, aggregated = evaluation.evaluate(qrels_2017, run)

aggregated

RUN: imi_mug_abs1 TOPICS: 30


{'recall_1000': 0.5366, 'ndcg': 0.4731, 'Rprec': 0.2392, 'P_10': 0.4467}

Run on 2018 topics and export to file for submission

In [9]:
run, run_params = running.run(topics_2018, 'ABSTRACTS', run_params)
run

RUN: imi_mug_abs1 TOPICS: 50


TOPIC_NO  Q0        ID  RANK      SCORE      RUN_NAME  \
0             1  Q0  23403819     1  60.751730  imi_mug_abs1   
1             1  Q0  25442222     2  58.318170  imi_mug_abs1   
2             1  Q0  27928645     3  58.140205  imi_mug_abs1   
3             1  Q0  22535154     4  58.069275  imi_mug_abs1   
4             1  Q0  24535907     5  57.269512  imi_mug_abs1   
5             1  Q0  22189819     6  56.515625  imi_mug_abs1   
6             1  Q0  26924424     7  56.431843  imi_mug_abs1   
7             1  Q0  25607474     8  56.346440  imi_mug_abs1   
8             1  Q0  27793752     9  56.277313  imi_mug_abs1   
9             1  Q0  29926184    10  55.992886  imi_mug_abs1   
10            1  Q0  25117819    11  55.913326  imi_mug_abs1   
11            1  Q0  23026937    12  55.673440  imi_mug_abs1   
12            1  Q0  23174497    13  55.335686  imi_mug_abs1   
13            1  Q0  17409425    14  55.294662  imi_mug_abs1   
14            1  Q0  23802768    15  55.290350  imi_mug_abs1   
15            1  Q0  25848750    16  55.195225  imi_mug_abs1   
16            1  Q0  22850568    17  55.088710  imi_mug_abs1   
17            1  Q0  29202777    18  55.053814  imi_mug_abs1   
18            1  Q0  25960652    19  54.957108  imi_mug_abs1   
19            1  Q0  27210749    20  54.920666  imi_mug_abs1   
20            1  Q0  26451873    21  54.848938  imi_mug_abs1   
21            1  Q0  26138035    22  54.808975  imi_mug_abs1   
22            1  Q0  24471189    23  54.595707  imi_mug_abs1   
23            1  Q0  27320919    24  54.535360  imi_mug_abs1   
24            1  Q0  27009410    25  54.397243  imi_mug_abs1   
25            1  Q0  25543407    26  54.277750  imi_mug_abs1   
26            1  Q0  26600396    27  54.236490  imi_mug_abs1   
27            1  Q0  16410717    28  54.222885  imi_mug_abs1   
28            1  Q0  25472943    29  54.078133  imi_mug_abs1   
29            1  Q0  24023633    30  54.012630  imi_mug_abs1   
...         ...  ..       ...   ...        ...           ...   
41362        50  Q0  18785507   971  45.581410  imi_mug_abs1   
41363        50  Q0  10698507   972  45.580666  imi_mug_abs1   
41364        50  Q0  27040395   973  45.568320  imi_mug_abs1   
41365        50  Q0  10720135   974  45.561530  imi_mug_abs1   
41366        50  Q0  29090521   975  45.549892  imi_mug_abs1   
41367        50  Q0  26524018   976  45.532616  imi_mug_abs1   
41368        50  Q0  27462774   977  45.528687  imi_mug_abs1   
41369        50  Q0  23713453   978  45.497395  imi_mug_abs1   
41370        50  Q0  25441922   979  45.496070  imi_mug_abs1   
41371        50  Q0  21079611   980  45.487724  imi_mug_abs1   
41372        50  Q0  12854887   981  45.475037  imi_mug_abs1   
41373        50  Q0  16627249   982  45.457996  imi_mug_abs1   
41374        50  Q0  24849514   983  45.450570  imi_mug_abs1   
41375        50  Q0  26375248   984  45.444360  imi_mug_abs1   
41376        50  Q0  12791658   985  45.437630  imi_mug_abs1   
41377        50  Q0  21182488   986  45.428978  imi_mug_abs1   
41378        50  Q0  28799432   987  45.427765  imi_mug_abs1   
41379        50  Q0  23382784   988  45.417255  imi_mug_abs1   
41380        50  Q0  29209600   989  45.412865  imi_mug_abs1   
41381        50  Q0  29907544   990  45.405476  imi_mug_abs1   
41382        50  Q0  22514634   991  45.402893  imi_mug_abs1   
41383        50  Q0  18490517   992  45.379967  imi_mug_abs1   
41384        50  Q0  23840454   993  45.370087  imi_mug_abs1   
41385        50  Q0  18729193   994  45.354027  imi_mug_abs1   
41386        50  Q0  23449638   995  45.352250  imi_mug_abs1   
41387        50  Q0  24065765   996  45.327500  imi_mug_abs1   
41388        50  Q0  17169806   997  45.322920  imi_mug_abs1   
41389        50  Q0  22411871   998  45.307697  imi_mug_abs1   
41390        50  Q0  20803351   999  45.274963  imi_mug_abs1   
41391        50  Q0  21324374  1000  45.246376  imi_mug_abs1   

                                      

In [10]:
utils.to_trec_run_file(run, run_params)

## SUBMITTED RUN 2

Baseline query with MeSH age groups and grid-search optimized parameters. Boost English and extra topics.
Relevance: 2

Evaluate on 2017 topics and qrels first.

In [11]:
run_params = {
    'run_id':'imi_mug_abs2',
    'query_template':'submitted/imi_mug_abs2.json',
}

run, run_params = running.run(topics_2017, 'ABSTRACTS', run_params)
results, aggregated = evaluation.evaluate(qrels_2017, run)

aggregated

RUN: imi_mug_abs2 TOPICS: 30


{'recall_1000': 0.5363, 'ndcg': 0.473, 'Rprec': 0.2391, 'P_10': 0.4467}

Run on 2018 topics and export to file for submission

In [12]:
run, run_params = running.run(topics_2018, 'ABSTRACTS', run_params)
run

RUN: imi_mug_abs2 TOPICS: 50


TOPIC_NO  Q0        ID  RANK      SCORE      RUN_NAME  \
0             1  Q0  23403819     1  60.928802  imi_mug_abs2   
1             1  Q0  25442222     2  58.495820  imi_mug_abs2   
2             1  Q0  27928645     3  58.317856  imi_mug_abs2   
3             1  Q0  22535154     4  58.246464  imi_mug_abs2   
4             1  Q0  24535907     5  57.446700  imi_mug_abs2   
5             1  Q0  22189819     6  56.693275  imi_mug_abs2   
6             1  Q0  26924424     7  56.609493  imi_mug_abs2   
7             1  Q0  25607474     8  56.523890  imi_mug_abs2   
8             1  Q0  27793752     9  56.454964  imi_mug_abs2   
9             1  Q0  29926184    10  56.169956  imi_mug_abs2   
10            1  Q0  25117819    11  56.090775  imi_mug_abs2   
11            1  Q0  23026937    12  55.851090  imi_mug_abs2   
12            1  Q0  23174497    13  55.513134  imi_mug_abs2   
13            1  Q0  17409425    14  55.472110  imi_mug_abs2   
14            1  Q0  23802768    15  55.467540  imi_mug_abs2   
15            1  Q0  25848750    16  55.372875  imi_mug_abs2   
16            1  Q0  22850568    17  55.266160  imi_mug_abs2   
17            1  Q0  29202777    18  55.231000  imi_mug_abs2   
18            1  Q0  25960652    19  55.134180  imi_mug_abs2   
19            1  Q0  27210749    20  55.098064  imi_mug_abs2   
20            1  Q0  26451873    21  55.026390  imi_mug_abs2   
21            1  Q0  26138035    22  54.986423  imi_mug_abs2   
22            1  Q0  24471189    23  54.773160  imi_mug_abs2   
23            1  Q0  27320919    24  54.712430  imi_mug_abs2   
24            1  Q0  27009410    25  54.574642  imi_mug_abs2   
25            1  Q0  26600396    26  54.413560  imi_mug_abs2   
26            1  Q0  16410717    27  54.399952  imi_mug_abs2   
27            1  Q0  25543407    28  54.277750  imi_mug_abs2   
28            1  Q0  25472943    29  54.255320  imi_mug_abs2   
29            1  Q0  24023633    30  54.190280  imi_mug_abs2   
...         ...  ..       ...   ...        ...           ...   
41362        50  Q0  10698507   971  45.757736  imi_mug_abs2   
41363        50  Q0  27040395   972  45.745773  imi_mug_abs2   
41364        50  Q0  10720135   973  45.738980  imi_mug_abs2   
41365        50  Q0  29090521   974  45.726963  imi_mug_abs2   
41366        50  Q0  27462774   975  45.706140  imi_mug_abs2   
41367        50  Q0  23713453   976  45.675045  imi_mug_abs2   
41368        50  Q0  25441922   977  45.673140  imi_mug_abs2   
41369        50  Q0  21079611   978  45.665173  imi_mug_abs2   
41370        50  Q0  12854887   979  45.652225  imi_mug_abs2   
41371        50  Q0  16627249   980  45.635445  imi_mug_abs2   
41372        50  Q0  24849514   981  45.627758  imi_mug_abs2   
41373        50  Q0  26375248   982  45.621544  imi_mug_abs2   
41374        50  Q0  12791658   983  45.615078  imi_mug_abs2   
41375        50  Q0  21182488   984  45.606426  imi_mug_abs2   
41376        50  Q0  28799432   985  45.605415  imi_mug_abs2   
41377        50  Q0  23382784   986  45.594704  imi_mug_abs2   
41378        50  Q0  29209600   987  45.590313  imi_mug_abs2   
41379        50  Q0  29907544   988  45.582874  imi_mug_abs2   
41380        50  Q0  18785507   989  45.581410  imi_mug_abs2   
41381        50  Q0  22514634   990  45.580290  imi_mug_abs2   
41382        50  Q0  18490517   991  45.557156  imi_mug_abs2   
41383        50  Q0  23840454   992  45.547535  imi_mug_abs2   
41384        50  Q0  26524018   993  45.532616  imi_mug_abs2   
41385        50  Q0  18729193   994  45.531425  imi_mug_abs2   
41386        50  Q0  23449638   995  45.529900  imi_mug_abs2   
41387        50  Q0  24065765   996  45.505150  imi_mug_abs2   
41388        50  Q0  17169806   997  45.500320  imi_mug_abs2   
41389        50  Q0  22411871   998  45.484886  imi_mug_abs2   
41390        50  Q0  20803351   999  45.452415  imi_mug_abs2   
41391        50  Q0  21324374  1000  45.423447  imi_mug_abs2   

                                      

In [13]:
utils.to_trec_run_file(run, run_params)

## SUBMITTED RUN 3

Baseline query with MeSH age groups and grid-search optimized parameters. Boost English and extra topics. Search age group and sex in title, abstract, and MeSH.
Relevance: 4

Evaluate on 2017 topics and qrels first.

In [14]:
run_params = {
    'run_id':'imi_mug_abs3',
    'query_template':'submitted/imi_mug_abs3.json',
}

run, run_params = running.run(topics_2017, 'ABSTRACTS', run_params)
results, aggregated = evaluation.evaluate(qrels_2017, run)

aggregated

RUN: imi_mug_abs3 TOPICS: 30


KeyboardInterrupt: 

Run on 2018 topics and export to file for submission

In [ ]:
run, run_params = running.run(topics_2018, 'ABSTRACTS', run_params)
run

In [ ]:
utils.to_trec_run_file(run, run_params)

## SUBMITTED RUN 4

Baseline query with MeSH age groups and grid-search optimized parameters. Boost English and extra topics. Search age group and sex in title, abstract, and MeSH. Abstract should exist.
Relevance: 5

Evaluate on 2017 topics and qrels first.

In [ ]:
run_params = {
    'run_id':'imi_mug_abs4',
    'query_template':'submitted/imi_mug_abs4.json',
}

run, run_params = running.run(topics_2017, 'ABSTRACTS', run_params)
results, aggregated = evaluation.evaluate(qrels_2017, run)

aggregated

Run on 2018 topics and export to file for submission

In [ ]:
run, run_params = running.run(topics_2018, 'ABSTRACTS', run_params)
run

In [ ]:
utils.to_trec_run_file(run, run_params)

## SUBMITTED RUN 5

Baseline query with MeSH age groups and grid-search optimized parameters. Boost English and extra topics. Abstract should exist.
Relevance: 1

Evaluate on 2017 topics and qrels first.

In [ ]:
run_params = {
    'run_id':'imi_mug_abs5',
    'query_template':'submitted/imi_mug_abs5.json',
}

run, run_params = running.run(topics_2017, 'ABSTRACTS', run_params)
results, aggregated = evaluation.evaluate(qrels_2017, run)

aggregated

Run on 2018 topics and export to file for submission

In [ ]:
run, run_params = running.run(topics_2018, 'ABSTRACTS', run_params)
run

In [ ]:
utils.to_trec_run_file(run, run_params)